feat(minifier): fold array and object constructors

camchenry opened the pull request:
This will fold expressions like `new Object()` to `{}`, and `new Array()` to `[]`. Based on the closure compiler tests: https://github.com/google/closure-compiler/blob/b7e380b6320a52ce7fbac0334ede00d6a8c92fad/test/com/google/javascript/jscomp/PeepholeSubstituteAlternateSyntaxTest.java#L78.

This is outside my usual area, so feedback is welcome.

NOTE: this was previously a full stack of PRs, but Graphite decided to stop working completely for some reason and only gave me this error when I submitted a PR:
```
ERROR: Failed to submit PR for 10-02-feat_minifier_fold_single_arg_new_array_expressions:
{}
```
so I decided to just completely remake this stack and submit as 1 PR.

camchenry commented:
* **#6257** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6257?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camchenry and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -323,6 +330,154 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+                    // `new Array()` -> `Array()`
+                    else {
+                        *expr = self.array_constructor_call(
+                            new_expr.arguments.clone_in(ctx.ast.allocator),
```
You can use `ctx.ast.move_vec`, since the previous one can be replaced. Using `clone_in` is performance-critical.

camchenry replied:
ah I knew there must be a way to do this, thank you!

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -89,6 +91,11 @@ impl<'a> Traverse<'a> for PeepholeSubstituteAlternateSyntax {
     }
 
+    fn exit_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        self.compress_new_expression(expr, ctx);
+        self.compress_call_expression(expr, ctx);
```
I recommend using the `match expr` pattern, and then pass `NewExpression<‘a>`, etc. type to the compressor function with the return `Expression<'a>`. Subsequently, replace these objects in the `Traverse` function. This approach allows us to align the code style of the replacement, enhancing the code’s sustainability.

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -323,6 +330,154 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+                    }
+                    // `new Array(8)` -> `Array(8)`
+                    else if arg.is_number_literal() {
+                        *expr = self.array_constructor_call(
+                            new_expr.arguments.clone_in(ctx.ast.allocator),
+                            ctx,
+                        );
+                        self.changed = true;
```
In this context, you can simply use `Some(Expression)` for changing (otherwise `None`) and replace it in `Traverse` impl. This provides a more convenient approach.

7086cmd commented on tasks/minsize/minsize.snap:
```diff
@@ -1,26 +1,26 @@
 544.10 kB  | 74.12 kB   | 72.48 kB   | 26.22 kB   | 26.20 kB   | lodash.js 
 
-555.77 kB  | 278.70 kB  | 270.13 kB  | 91.39 kB   | 90.80 kB   | d3.js     
+555.77 kB  | 278.24 kB  | 270.13 kB  | 91.36 kB   | 90.80 kB   | d3.js     
```
Excellent improvement, btw! 🎊

camchenry commented:
@7086cmd Thanks for the feedback, I updated the PR and I think it's a little bit simpler and more consistent now. I wasn't sure how to avoid the `(*arg).clone_in()` part though, so if you know a trick for that I'd love to hear it.

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -327,6 +347,143 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+            if new_expr.arguments.is_empty() {
+                Some(self.empty_array_literal(ctx))
+            } else if new_expr.arguments.len() == 1 {
+                let Some(arg) = new_expr.arguments[0].as_expression() else { return None };
```
```suggestion
                let Some(arg) = new_expr.arguments.get_mut(0).and_then(|a| a.as_expression_mut())
                else {
                    return None;
                };
```
Use the `_mut()` method, which can finally return `&mut Expression`.

7086cmd replied:
In fact, the `get_mut` function may not be necessary in this scenario. However, we can utilize it to prevent panic in this particular situation.

camchenry replied:
I like the `get_mut` here 👍

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -327,6 +347,143 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+                // `new Array(literal)` -> `[literal]`
+                else if arg.is_literal() || matches!(arg, Expression::ArrayExpression(_)) {
+                    let mut elements = Vec::new_in(ctx.ast.allocator);
+                    let element = ctx
+                        .ast
+                        .array_expression_element_expression((*arg).clone_in(ctx.ast.allocator));
```
```suggestion
                    let element =
                        ctx.ast.array_expression_element_expression(ctx.ast.move_expression(arg));
```

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -327,6 +347,143 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+                // `new Array(1, 2, 3)` -> `[1, 2, 3]`
+                let elements = Vec::from_iter_in(
+                    new_expr.arguments.iter().filter_map(|arg| arg.as_expression()).map(|arg| {
+                        ctx.ast
+                            .array_expression_element_expression((*arg).clone_in(ctx.ast.allocator))
```
The rest are similar, so you can manually handle it.

7086cmd reviewed (dismissed):
Just let the reference mutable, so you can directly move the expression.

7086cmd commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -327,6 +347,145 @@ impl<'a> PeepholeSubstituteAlternateSyntax {
+        if new_expr.arguments.is_empty()
+            && new_expr.callee.is_global_reference_name("Object", ctx.symbols())
+        {
+            Some(ctx.ast.expression_object(new_expr.span, Vec::new_in(ctx.ast.allocator), None))
```
Use `ctx.ast.vec()` is more visible.

camchenry committed:
feat(minifier): fold array and object constructors (#6257)

This will fold expressions like `new Object()` to `{}`, and `new Array()` to `[]`. Based on the closure compiler tests: https://github.com/google/closure-compiler/blob/b7e380b6320a52ce7fbac0334ede00d6a8c92fad/test/com/google/javascript/jscomp/PeepholeSubstituteAlternateSyntaxTest.java#L78.

This is outside my usual area, so feedback is welcome.

NOTE: this was previously a full stack of PRs, but Graphite decided to stop working completely for some reason and only gave me this error when I submitted a PR:
```
ERROR: Failed to submit PR for 10-02-feat_minifier_fold_single_arg_new_array_expressions:
{}
```
so I decided to just completely remake this stack and submit as 1 PR.
